multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,201 @@
+      multiLineRequest.setAddressNumber(verifyAddress.getAddressNumber)
+      multiLineRequest.setStreet(verifyAddress.getStreet)
+      multiLineRequest.setAdmin1(verifyAddress.getAdmin1.getLongName)
+      multiLineRequest.setAdmin2(verifyAddress.getAdmin2.getLongName)
+      multiLineRequest.setCity(verifyAddress.getCity.getLongName)
+      multiLineRequest.setPostalCode(verifyAddress.getPostalCode)
+      multiLineRequest.setCountry(verifyAddress.getCountry.getIsoAlpha3Code)
+
+    }
+
+    // Perform the geocode and return the result if it passes our acceptance criteria
+    val multiLineResponse = addressing.geocode(multiLineRequest, relaxedGeocoderPreferences)
+    if (CollectionUtils.isNotEmpty(multiLineResponse.getResults) && isAddressLevelMatch(multiLineResponse.getResults.get(0))) {
+      return multiLineResponse
+    }
+
+    // build a single line version of the address for the second pass of geocoding, based on the verified address or the input
+    val singleLineRequest: RequestAddress = new RequestAddress()
+    if (CollectionUtils.isNotEmpty(verifyResponse.getResults)) {
+      val verifyAddress = verifyResponse.getResults.get(0).getAddress
+      singleLineRequest.setAddressLines(Collections.singletonList(verifyAddress.getFormattedAddress))
+      singleLineRequest.setCountry(verifyAddress.getCountry.getIsoAlpha3Code)
+    } else {
+      singleLineRequest.setAddressLines(Collections.singletonList(
+        inputRequest.getStreet + " " +
+          inputRequest.getCity + " " +
+          inputRequest.getAdmin1 + " " +
+          inputRequest.getPostalCode
+      ))
+      singleLineRequest.setCountry(inputRequest.getCountry)
+    }
+
+    // Perform the single line geocode and return the result if it passes our acceptance criteria
+    val singleLineResponse = addressing.geocode(singleLineRequest, relaxedGeocoderPreferences)
+    if (CollectionUtils.isNotEmpty(singleLineResponse.getResults) && isAddressLevelMatch(singleLineResponse.getResults.get(0))) {
+      return singleLineResponse
+    }
+
+    // neither address passed our initial acceptance criteria... pick the best response
+    if (CollectionUtils.isNotEmpty(multiLineResponse.getResults)) {
+      // Return the single line response only if it is better than the multi line response
+      if (CollectionUtils.isNotEmpty(singleLineResponse.getResults)
+        && singleLineResponse.getResults.get(0).getScore > multiLineResponse.getResults.get(0).getScore) {
+        return singleLineResponse
+      }
+      multiLineResponse
+    } else {
+      singleLineResponse
+    }
+  }
+
+  def isAddressLevelMatch(result: Result): Boolean =
+    result.getScore >= 90 && "ADDRESS".equals(result.getExplanation.getAddressMatch.getType.label)
+
+  def getMainAddressLines(addressLines: List[String]): List[String] = {
+    val mainAddressLine = new ArrayList[String]()
+    mainAddressLine.addAll(addressLines.dropRight(1).filter(_.nonEmpty))
+    mainAddressLine
+  }
```

Comment:
why is the "addAll" needed?  Can't this be more simply:
```
def getMainAddressLines(addressLines: List[String]): List[String] = {
    addressLines.filter(_.nonEmpty).dropRight(1)
  }
```
Also - i think it might be important to filter before dropping - in case somehow they had extra blanks added in the output lines field.

Reply:
Yes, earlier I tried it in above mentioned way and for sample code, it is working fine. However, when we return this filtered addressLines list and pass to multiLine request, geocoder is giving null values. So, I tried with new List and it is working fine in this way. I don't know what is going behind in scala at the time of returning the List[string] from addressLines.dropRight(1).filter(_.nonEmpty).

Reply:
For filter before dropping, 

My thought was If we are getting addressNumber, street in 1st line, & city/state/postal in second and if second line comes empty. then it will give us empty list if we are using dropRight(1) before filter although we have addressNumber, street in it.

but yeah, I agree with you so I will make changes accordingly